web-console/src/lib/compositions/streaming/management/usePipelineMetrics.ts

```diff
@@ -1,74 +1,62 @@
 import { usePipelineManagerQuery } from '$lib/compositions/usePipelineManagerQuery'
 import { nonNull } from '$lib/functions/common/function'
-import {
-  ConnectorStatus,
-  GlobalMetrics,
-  InputConnectorMetrics,
-  OutputConnectorMetrics,
-  PipelineStatus
-} from '$lib/types/pipeline'
-import { useEffect, useState } from 'react'
+import { tuple } from '$lib/functions/common/tuple'
+import { ControllerStatus, InputEndpointMetrics, OutputEndpointMetrics, PipelineStatus } from '$lib/types/pipeline'
+import { useState } from 'react'
 
 import { useQuery } from '@tanstack/react-query'
 
+const toMetrics = (data: ControllerStatus, { timeMs }: { timeMs: number }) => ({
+  input: new Map(data.inputs.map(cs => tuple(cs.config.stream, cs.metrics))),
+  output: new Map(data.outputs.map(cs => tuple(cs.config.stream, cs.metrics))),
+  global: [{ ...data.global_metrics, timeMs }]
+})
+
+const emptyData = {
+  input: new Map<string, InputEndpointMetrics>(),
+  output: new Map<string, OutputEndpointMetrics>(),
+  global: []
+}
+
 export function usePipelineMetrics(props: {
   pipelineName: string
   status: PipelineStatus
   refetchMs: number
   keepMs?: number
 }) {
-  const [metrics, setMetrics] = useState({
-    global: [] as GlobalMetrics[],
-    input: new Map<string, InputConnectorMetrics>(),
-    output: new Map<string, OutputConnectorMetrics>()
-  })
   const pipelineManagerQuery = usePipelineManagerQuery()
+  const [lastTimestamp, setLastTimestamp] = useState<number>()
   const pipelineStatsQuery = useQuery({
     ...pipelineManagerQuery.pipelineStats(props.pipelineName),
-    enabled: props.status == PipelineStatus.RUNNING,
-    refetchInterval: props.refetchMs
-  })
+    enabled: props.status === PipelineStatus.RUNNING,
+    refetchInterval: props.refetchMs,
+    refetchIntervalInBackground: true,
+    refetchOnWindowFocus: false,
+    structuralSharing: (oldData: any, _newData: any) => {
+      if (!_newData) {
+        setLastTimestamp(undefined)
+        return emptyData
+      }
+      const now = Date.now()
+      if (!lastTimestamp) {
+        setLastTimestamp(now)
+      }
+      const newData = toMetrics(_newData, { timeMs: lastTimestamp ? now - lastTimestamp : 0 })
+      const oldGlobal = oldData?.global ?? []
```

Comment:
I don't think you should get the timestamps from the UI, they should come with the data.

Reply:
They don't yet =)